README
ReSCAL: a Real-Space Cellular Automaton Laboratory
--------------------------------------------------
Copyright (C) 2011-2016


Description
-----------

This is ReSCAL, a cellular automaton modeling tool in real 3d space, 
primarily designed for geomorphology. However we believe it is 
sufficiently generic in its approach to deal with a wide range of 
applications. 
Basically, it simulates asynchronous first-neighbour interactions 
(see the papers in the References section). We refer to Markov 
chains and Poisson processes for the evolution of the physical time.

We also implemented a number of additional mechanisms that are 
relevant in usual geophysical models
- correlations (coupled processes)
- avalanches (granular processes)
- lattice gas coupling (erosion)
- rotating table (multidirectional winds)
- map surrections (tectonic processes)


Content
-------

ReSCAL package comprises several programs:
- rescal: ReSCAL main application
- genesis: generation of a 3d cellular space (CSP or raw binary format)
- regenesis: modification of a cellular space
- csp2png: PNG image generator from CSP-file
- bin2png: PNG image generator from BIN-file
- cspinfo: CSP metadata reader
- csp2bin: CSP to raw binary conversion tool

Samples are present in the scripts directory, along with Matlab/octave
functions that support CSP format.


Requirements
------------

Needed libraries are glib-2.0, gtk+-2.0, gdk-pixbuf and pthread.
'genesis.py' uses NumPy, but it is possible to use C version of 
genesis instead.


Getting started
---------------

1) Build all programs by typing
./configure
make

(See also 'INSTALL' file for detailled instructions)

2) Go to the scripts directory and launch the 'run' script:
cd scripts
./run

This will open a window, with no menu, displaying a dunes simulation 
controlled by air flow. Note the initial phase of flow stabilization.

The environment variable OMP_NUM_THREADS lets you set the number of
OpenMP threads in the computation of the flow. You can get optimal 
speed by setting this variable according to the number of available 
CPU cores.

(See rescal command line options to customize display preferences
in the script)

3) Edit the 'run' script and change the value of PAR_FILE:
PAR_FILE="dun_rot.par"
Then launch rescal again:
./run

Now you have a similar execution, with rotations occuring regularly,
as if the wind was rotating. Again the flow needs to be stabilized
after each rotation.

4) Edit a parameter file, change some values and see what happens in 
your model.

For example, you can change the overall size by editing the 'H', 'L', 
'D' parameters in 'dun.par'.
H = heigth (vertical)
L = length (from east to west)
D = depth (from north to south)

Beware that a huge cellular space will slow down the simulation.

Another suggestion would be to choose a predefined CSP template for
the initial configuration of the cellular space. For example, you can 
get a conical pile of sand by setting:
Csp_template = RCONE 

5) Save the output files of your last simulation by using the 'xrep' 
script.


Advanced usage
--------------

It is possible to select and/or modify any of the predefined models.
So far, this requires to edit some of the sources in C language. 
Here we briefly explain how to proceed. Ideally, we assume 
that the user have some notion of programming.

1) Edition of genesis sources

The genesis program is provided in python and C languages. These two
standalone versions are respectively located in 'scripts' and 
'src' directories. The python version of genesis is easier to
read, however it does not support the CSP format yet, nor any of the
CSP templates.

Note that for historical reasons, a cellular space is composed of 
vertical plans in the east-west direction. Each plan is a 
concatenation of horizontal lines from left to right. The very first 
cell corresponds to the north-upper-left corner.

When lattice gas is used, a floor and a ceiling of DUM cells are 
required to prevent gas particles from escaping the cellular space.

2) Edition of rescal sources

Start by selecting a model and its specific options, in 'defs.h'. 

You may also edit the transitions in 'models.c'. Basically, the 
syntax looks like

trans(direction, cel1_start, cel2_start, cel1_end, cel2_end, rate).

Then you will get an oriented transition 

[cel1_start, cel2_start] -> [cel1_end, cel2_end] 

occuring at the specified rate. See USER'S GUIDE in 'models.c' for 
an overview of the available functions.

Use the command-line option '-info' to generate a number of log files.
For instance, the list of all transitions is present in 'TRANS.log'.

Feel free to have a look around and play with the existing mechanisms
in 'space.c' and 'surface.c', or to suggest any improvement you would 
like to see. Obviously, there is still a lot left to do ! (See the 
TODO file)

A modified version can be freely distributed provided that you keep 
intact all the notices that refer to the license (See 'COPYING' file).

We would be grateful if you mention the use of this program in a 
related work or publication.


References
----------

[1] On a small scale roughness of the core-mantle boundary
C. Narteau, J.L. Le Mouël, J.P. Poirier, E. Sepulveda and M. Shnirman
Earth and Planetary Science Letters, 191, 49-60 (2001).

[2] Setting the length and time scales of a cellular automaton dune 
model from the analysis of superimposed bedforms
C. Narteau, D. Zhang, O. Rozier and P. Claudin
Journal Geophysical Research, 114, F03006 (2009).

[3] Morphodynamics of barchan and transverse dunes using a cellular 
automaton model
D. Zhang, C. Narteau and O. Rozier
Journal Geophysical Research, 115, F03041 (2010).


========================================================================

ReSCAL is free software; you can redistribute it and/or modify it under the
terms of the GNU General Public License as published by the Free Software
Foundation. It is distributed in the hope that it will be useful, but WITHOUT
ANY WARRANTY WHATSOEVER.

Check the License for details. You should have received a copy of it, along
with the package; see the file 'COPYING'. If not, write to the Free Software
Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.




NEWS
Version 1.6
-----------
AVX optimization of lattice-gas propagation.
Optimization of flow interpolation.
New command-line option (-pat) for the concurrent execution of stochastic engine 
and lattice gas, with non-deterministic algorithm.
Command-line help improved with new options -h and -hm.
More CSP templates and parameters for DUN model.
Various fixes.

Version 1.5
-----------
Relief light-shading with various colors, for colored sand and topographic 
obstacles (DUM cells) in DUN and AVA models, and for fluid cells in RIV model.
More simple syntax for graphical options in command-line.
The graphical display readapts automatically as the window gets larger.
Better interactivy with the positioning of cross sections.
New TRACE_FLUX mode in DUN model for the calculation of the average sand flux
over time, compatible with rotations.
New option for stability analysis in DUN model.
New parameters and transtions for the mobility of sand in DUN model.
Introduction of vegetated cells in the DUN model, with specific transitions
and parameters.
Time metadata added in CSP format.
More CSP templates, and arguments added in previous templates.
Many tweaks and fixes.

Version 1.4
-----------
New FULL rotation mode for rotating the whole cellular space with support of
periodic boundary conditions and lattice gas coupling.
More CSP templates and samples.
RIV model improved.
More graphical options and parameters.
Scripts added for video encoding.
Various fixes and optimizations.

Version 1.3.2
-------------
OpenMP support in lgca module.
New trace mode TRACE_PAR_COL for the tracking of colored particles.

Version 1.3.1
-------------
New parameter for boundary conditions.
Library gd not needed anymore.

Version 1.3
-----------
The time scale is calculated in DUN model, and t0 time unit can be 
used for all delays on the command-line.
Cellular space is automatically reoriented on images, after rotations.
New parameters with CELL_COLOR option.
Support for CSP templates in genesis.
New LIFE model implemented.

Version 1.2
-----------
CSP format with metadata has been implemented.
New parameters available for avalanches and lattice gaz.
Interactive display and new disposition in drawing area.

Version 1.1
-----------
Integration of glade interface with toolbar and statusbar.
Compatibility with gtk+ 1.x has been abandoned.

Version 1.0
-----------
This is the first release of ReSCAL, under the GNU General Public 
License.  It is far from complete. But it provides a number of 
predefined models with some parameters and mechanisms, included 
lattice gaz coupling, and real-time display with several graphical 
options.



src/defs.h
/* ReSCAL - Global definitions
 *
 * Copyright (C) 2011-2016
 *
 * Author: Olivier Rozier <rozier@ipgp.fr>
 *
 * This file is part of ReSCAL.
 *
 * ReSCAL is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 */

//#ifdef HAVE_CONFIG_H
//#  include <config.h>
//#endif

/// version number and date
#define VER_NUM "1.6"
#define VER_DAT "2016-10"

/// numeric format
#define NUM_MODE "C"  //C language
//#define NUM_MODE "en_US.UTF-8"  //english
//#define NUM_MODE "fr_FR.UTF-8"  //french

/// ======================================================
/// programs
enum PROGS {PROG_RESCAL, PROG_TOOL};

/// ======================================================

/// choice of a model
//#define MODEL_DUN    //Dune morphodynamics
//#define MODEL_AVA    //Avalanches
//#define MODEL_CMB    //Core-mantle boundary
//#define MODEL_ICB    //Innercore-core boundary
//#define MODEL_RIV    //River (landscape evolution)
//#define MODEL_CRY    //Crystallization (of calcite)
//#define MODEL_DIF    //Diffusion
//#define MODEL_D2G    //Diffusion 2d with gravity
//#define MODEL_LIFE   //Stochastic game of life (2D)
#define MODEL_SNO    //Snow bedform morphodynamics

/// ======================================================

/// common options
// #define OPENMP		//use OpenMP parallelization (on lattice gas)
#define GUI		//graphical interface
#define CYCLAGE_HOR     //enable horizontal cycling
#define INFO_CEL        //log the number of cells
#define INFO_DBL        //log the number of active doublets
#define INFO_TRANS      //log the number of effective transitions
#define DETERMINISTIC	  //try to keep a deterministic execution: same parameters produce same results (it may be slower when DETERMINISTIC is defined, but not always !)
#define DUMP_SIGNATURE  //footprints for the comparison of two simulations (identical simulations => identical footprints)
//#define LOG_FILE        //ReSCAL logs redirected into a file
#define USE_LIBPNG        //use of libpng
//#define USE_GD          //use of libgd (alternative to USE_LIBPNG, but obsolete)
//#define ROTATE_LIGHT    //automatic rotation of the light source in ROT_CYCLE mode
#define REORIENT_AUTO   //automatic reorientation when saving (images of) the cellular space
#define CSP_MUTEX       //mutual exclusion on the cellular space (no transition while saving data)
//#define GTK_OLD_SYNTAX  //compile with old version of GTK+2.x (obsolete)
// #define PARALLEL_AUTOMATA //enable concurrent execution of stochastic engine and lattice gas  (non-deterministic only, not working on MacOS X)


#if !defined(_OPENMP) || defined(__CYGWIN32__)
#undef OPENMP
#endif

#if !defined(USE_AVX) && defined(__AVX__)
#define USE_AVX
#endif

/// ===================== CMB MODEL =====================
#ifdef MODEL_CMB

/// name of the model
#define MOD_NAME "CMB"

/// description
#define MOD_DESC "simulation of the core-mantle boundary"

/// cell states
#define PLUS  0 //solide
#define ZERO  1 //liquide sature en elements legers
#define MOINS 2 //liquide non-sature
#define BORD  3 //bord
#define DUM   4 //inerte
#define PIERRE 7 //granit
#define TUNNEL 9 //tunnel vers une terre parallele

/// names of cell states
#define ETATS { "PLUS", "ZERO", "MOINS", "BORD", "DUM", "", "", "GRANIT", "", "TUNNEL" }

/// phases of cell states
#define PHASES { SOLID, FLUID, FLUID, SOLID, SOLID, SOLID, SOLID, SOLID, SOLID, SOLID }

/// specific options of the model
#define DUMP_RUGOSI
//#define ROTATIONS
#endif

/// ===================== SNO MODEL =====================
// KK 10 May 2018
// Most cell types are inherited from the DUN MODEL (below)
#ifdef MODEL_SNO
/// name of the model
#define MOD_NAME "SNO"
#define MOD_DESC "simulation of snow surface under air flow"

// specific options of the model (more options available under DUN MODEL)
#define COHESION // enable some grains (GRV) to be harder to remove than others
#define SINTER   // time-dependent hardening - not yet implemented

#endif

/// ===================== DUN MODEL =====================
//  Most properties of dunes - except name and vegetation - are inherited by SNO MODEL
#ifdef MODEL_DUN

/// name of the model
#define MOD_NAME "DUN"

/// description
#define MOD_DESC "simulation of sand dunes under fluid flow"
#endif // MODEL_DUN

#if defined(MODEL_DUN) || defined(MODEL_SNO)
/// cell states
#define GR    0 //grain
#define GRJ   1 //mobile grain
//#define BR    2 //bedrock (no transitions)
#define GRV    2 //vegetated grain
#define EAUC  3 //air (or water)
#define VEG   4 //vegetation
#define BORD  5 //boundary
#define DUM   6 //neutral
#define IN    7 //input of sand
#define OUT   8 //output of sand
#define TUNNEL 9 //tunnel to another cellular space (with PARALLEL option)
#define EAUT  10 //(used only graphically, no transitions)
#define GRC   11 //colored grain (used only graphically, no transitions)

/// names of cell states
//#define ETATS { "GR", "GRJ", "BR", "EAUC", "VEG", "BORD", "DUM", "IN", "OUT", "TUNNEL" }
#define ETATS { "GR", "GRJ", "GRV", "EAUC", "VEG", "BORD", "DUM", "IN", "OUT", "TUNNEL" }

/// phases of cell states
#define PHASES { SOLID, SOLID, SOLID, FLUID, SOLID, SOLID, SOLID, SOLID, SOLID, SOLID }
//#define PHASES { SOLID, FLUID, SOLID, FLUID, SOLID, SOLID, SOLID, SOLID, SOLID, SOLID }

/// specific options of the model
#define ALTI GR //compute elevation map for cells of type GR
#define AVALANCHES
#define ROTATIONS
#define CENTERING_AUTO //automatic centering
#define LIENS_TRANSITIONS //correlations of transitions
#define LGCA //lattice gas
#define CGV //compute shear stress (with LGCA option)
#define WIND_DATA //option for importing wind data from a file (with ROTATIONS option)
#define TIME_SCALE //option for computing the time scale (using additional physical parameters)
//#define CELL_TIME GR //save time for each cell of type GR
//#define CELL_COLOR //use colored cells
//#define NORM2D //CGV with 2d normals
//#define TRACE_TRANS
//#define TRACE3D_CEL
//#define TRACE_PAR_COL //save the courses of colored cells (one cell per color)
//#define TRACE_FLUX //save the mean flux of grains (2d)
//#define STABILITY_ANALYSIS //produce data for the 2D linear stability analysis (find the length scale of the model)
//#define DUMP_SIGMA //computation of the standard deviation in height (when using WAVES_2D template)
//#define DUMP_AUTOCORREL //computation of the auto-correlation in height (when using WAVES_2D template)
//#define APPEL_MCC //external matlab application
//#define USE_VEGETATION //option for using cells in VEG state

#endif // DUN or SNO

/// ===================== AVA MODEL =====================
#ifdef MODEL_AVA

/// name of the model
#define MOD_NAME "AVA"

/// description
#define MOD_DESC "simulation of granular avalanches"

/// cell states
#define GR    0 //grain stable
#define AIR   1 //air
#define BORD  2 //bord
#define DUM   3 //inerte
#define IN    4 //source de grains
#define OUT   5 //sortie de grains
#define TUNNEL 9 //tunnel vers une terre parallele

#define TERRE    GR

/// names of cell states
#define ETATS { "GR", "AIR", "BORD", "DUM", "IN", "OUT", "", "", "", "TUNNEL" }

/// phases of cell states
#define PHASES { SOLID, FLUID, SOLID, SOLID, SOLID, SOLID }

/// specific options of the model
#define ALTI GR //calcul de la hauteur des piles de cellules GR
#define AVALANCHES  //avalanches
//#define ROTATIONS //table tournante
//#define CELL_TIME GR //datation des cellules GR
#define CELL_COLOR //cellules colorees
//#define NORM2D //pour l'affichage
//#define TRACE_TRANS
#endif

/// ===================== ICB MODEL =====================
#ifdef MODEL_ICB

/// name of the model
#define MOD_NAME "ICB"

/// description
#define MOD_DESC "simulation of the Earth's inner-core boundary"

/// cell states
#define PLUS  0 //fer solide
#define ZERO  1 //liquide sature en elements legers
#define MOINS 2 //liquide non-sature
#define BORD  3 //bord
#define DUM   4 //inerte
#define TUNNEL 9 //tunnel vers une terre parallele

/// names of cell states
#define ETATS { "PLUS", "ZERO", "MOINS", "BORD" , "DUM"}

/// phases of cell states
#define PHASES { SOLID, FLUID, FLUID, SOLID, SOLID }

/// specific options of the model
#define DUMP_RUGOSI

#endif

/// ===================== RIV MODEL =====================
#ifdef MODEL_RIV

/// name of the model
#define MOD_NAME "RIV"

/// description
#define MOD_DESC "simulation of erosion by runoff"

/// cell states
#define TERRE  0 //terre
#define BOUE   1 //boue
#define EAU    2 //eau
#define BT     3 //boue turbulent
#define PIERRE 4 //pierre
#define BORD   5 //bord
#define DUM    6 //inerte
#define IN     7 //source
#define OUT    8 //sortie
#define TUNNEL 9 //tunnel vers une terre parallele

/// names of cell states
#define ETATS { "TERRE", "BOUE", "EAU", "BT", "PIERRE", "BORD", "DUM", "IN", "OUT", "TUNNEL"}

/// phases of cell states
#define PHASES { SOLID, FLUID, FLUID, FLUID, SOLID, SOLID, SOLID, SOLID, SOLID, SOLID }

/// specific options of the model
#define LIENS_TRANSITIONS
#define ALTI TERRE //calcul de la hauteur des piles de cellules TERRE
#define AVALANCHES
#define SURRECTIONS
//#define TRACE_SRC //calcul bassin versant
//#define TRACE_AIRE  //calcul aires drainees
//#define TRACE_TRANS
//#define TRACE_PAR //calcul debit + pente
//#define APPEL_GMT
//#define LISSAGE  //couche uniforme de BOUE + cristallisation BOUE->TERRE + pas de transport vers le haut => lissage avant mode TRACE
//#define CELL_TIME TERRE //datation des cellules TERRE


#endif


/// ===================== CRYSTAL MODELE =====================
#ifdef MODEL_CRY

/// name of the model
#define MOD_NAME "CRYSTAL"

/// description
#define MOD_DESC "simulation of calcite crystal growth"

/// cell states
#define PLUS  0 //solide
#define ZERO  1 //gaz
#define BORD  3 //bord
#define DUM   4 //inerte
#define IN    5 //source du gaz
#define AIR   6 //air
#define TUNNEL 9 //tunnel vers une terre parallele

/// names of cell states
#define ETATS { "SOLIDE", "GAZ", "", "BORD", "DUM", "SOURCE", "AIR" }

/// phases of cell states
#define PHASES { SOLID, FLUID, SOLID, SOLID, SOLID, SOLID, FLUID }

#endif

/// ===================== DIF MODEL =====================
#ifdef MODEL_DIF

/// name of the model
#define MOD_NAME "DIF"

/// description
#define MOD_DESC "isotropic diffusion"

/// cell states
#define ZERO  0 //gaz 0
#define ONE   1 //gaz 1
#define IN    2 //source
#define BORD  3 //bord
#define DUM   4 //inerte


/// names of cell states
#define ETATS { "ZERO", "ONE", "IN", "BORD", "DUM" }

/// phases of cell states
#define PHASES { FLUID, FLUID, SOLID, SOLID, SOLID }

#endif

/// ===================== D2G MODEL =====================
#ifdef MODEL_D2G

/// name of the model
#define MOD_NAME "D2G"

/// description
#define MOD_DESC "diffusion 2d with gravity"

/// cell states
#define TERRE 0 //terre
#define EAU  1 //eau
#define AIR   2 //air
#define BORD  3 //bord
#define DUM   4 //inerte
#define IN    5 //source
#define BOUE  6 //boue

/// names of cell states
#define ETATS { "TERRE", "EAU", "AIR", "BORD", "DUM", "SOURCE", "BOUE" }

/// phases of cell states
#define PHASES { SOLID, FLUID, FLUID, SOLID, SOLID, SOLID, FLUID }

#endif

/// ===================== LIFE MODEL =====================
#ifdef MODEL_LIFE

/// name of the model
#define MOD_NAME "LIFE"

/// description
#define MOD_DESC "stochastic game of life (2D)"

/// cell states
#define DEAD 0
#define ALIVE 1
#define BORD  2 //bord
#define DUM   3 //inerte

/// names of cell states
#define ETATS { "DEAD", "ALIVE", "BORD", "DUM" }

#endif

/// ======================================================

/// verification of compatibility between the options

#ifndef ROTATIONS
#undef REORIENT_AUTO
#endif

#ifndef CGV
#undef TIME_SCALE
#endif

#ifndef CELL_COLOR
#undef TRACE_PAR_COL
#endif

#ifdef STABILITY_ANALYSIS
#define NORM2D
#define DUMP_SIGMA
#endif

#ifndef LGCA
#undef PARALLEL_AUTOMATA
#endif

//#ifdef PARALLEL_AUTOMATA
//#undef DETERMINISTIC
//#endif

/// ======================================================

/// common constants

#define MAX_CELL  12  //max number of cell types
#define MAX_DB    100 //max number of doublets in the model
#define MAX_CHK   3   //max number of check functions for one transition

#define EST       0
#define BAS       1
#define SUD       2

#define OUEST     3
#define HAUT      4
#define NORD      5

#define VERTICAL   0
#define EST_OUEST  1
#define NORD_SUD   2
#define HORIZONTAL 3 // EST_OUEST ou NORD_SUD
#define ISOTROPE   4
#define CLASSES_DB { "VERTICAL", "EAST_WEST", "NORTH_SOUTH", "HORIZONTAL" ,"ISOTROPIC"}

#define MAX_TRANSITIONS_DB 100
#define MAX_TRANSITIONS_CEL 10
#define MAX_LIENS  100

#define MAX_TRANSITIONS MAX_TRANSITIONS_DB+MAX_TRANSITIONS_CEL

#ifdef PARALLEL
#define MAX_NODE  16
#endif

#define PI 3.14159265

#define W32 1
#define W64 2

#define SOLID 0
#define FLUID 1

/// ======================================================

